Let Makefile handle assets extraction

yne opened the pull request:
Tested on multiple Linux and on MacOS

Because of the CRLF convertion, the changes are unreadable with whitespace. Better use
https://github.com/snesrev/zelda3/pull/31/files?w=1

snesrev commented:
Is it possible to invoke '/usr/bin/env python3' from the makefile instead to support CRLF in the .py files?

ZachBacon commented:
ideally could do something like PYTHON=path/to/python make

yne commented:
@ZachBacon Good Idea, I've added an `PYTHON=/path/to/python` variable that default to `/usr/bin/env python3`.

```sh
# example to override the default /usr/bin/env python3
make PYTHON=python3 clean all
```

@snesrev okay, I removed the shebang and LF conversion. all fine tuning (-Wall, -Os, -O3 -lto...) shall be done by the end user, not the makefile. hence the lack of -lto in Makefile

Also: consider avoiding `atan` as it add another lib dependency to the project (`-lm`) for just 1 function.

snesrev commented:
I merged this, but I had to add -O2 and -lm. Is there some way to add -O2 by default, but still making it possible to override on the command line?

yne commented:
none that I'm aware of.

Just out of curiosity, why is the `-O2` mandatory ?

If it was not, you could have added `CFLAGS=-O2 make` to the examples

snesrev commented:
I want the default to be -O2 because I want people to build optimized binaries so it uses less cpu by default.
Is there a way to put all the .o files in a separate directory?

snesrev commented:
People type make by habit.. so the default unless you don't change CFLAGS should be -O2.

yne commented:
Edit: maybe by placing the `-O2` *before* the CFLAGS expansion ?
maybe by placing the -O before the user-given CFLAGS

```diff
-CFLAGS:=${CFLAGS} -O2 $(shell sdl2-config --cflags)
+CFLAGS:=-O2 ${CFLAGS} $(shell sdl2-config --cflags)
```

So the cc command will be `-O2 -O3 ...`

GCC doesn't complain having multiple -O flags, but I'm unsure about other compiler

snesrev commented:
Perhaps something like this:
CFLAGS := $(if $(CFLAGS),$(CFLAGS),$(something else))

snesrev commented:
I pushed something that seems to work.

snesrev commented:
Make -j8 doesn't seem to work anymore though. It's related to building of resources. It's not waiting for the resources before it starts building .c files.

snesrev commented:
I don't think this will work because tables/generated_*.h don't exist the first time you run make if the autogenerated files don't exist yet.
GEN:=$(shell grep -hor tables/generated.*.h --include \*.c .)

yne commented:
Okay, we need to explicit the compilation rule to add it on of the generated ressources

```
%.o : %.c tables/generated_dialogue.h
    $(CC) -c $(CFLAGS) $< -o $@
```

yne commented:
An interesting addition would be to split the python extraction/compilation of each .h files into multiple invocation (instead of 2 big extract+compil monostep).

This way `make` could multithread the asset extraction/compilation and even detect which extracted assets needs to be re-compiled into .h.

But it's not that important.
